src/group_theory/solvable.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2020 Jordan Brown, Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jordan Brown, Thomas Browning and Patrick Lutz
+-/
+
+import group_theory.abelianization
+
+/-!
+# Solvable Groups
+
+In this file we introduce the notion of a solvable group. We define a solvable group as one whose
+nth commutator is `1` for some `n`. This requires defining the commutator of two subgroups and
+the derived series of a group.
+
+## Main definitions
+
+* `general_commutator H₁ H₂` : the commutator of the subgroups `H₁` and `H₂`
+* `derived_series G n` : the `n`th term in the derived series of `G`, defined by iterating
+  `general_commutator` starting with the top subgroup
+* `is_solvable G` : the group `G` is solvable
+-/
+
+open subgroup
+
+variables {G : Type*} [group G]
+
+section general_commutator
+
+/-- The commutator of two subgroups `H₁` and `H₂`. -/
+def general_commutator (H₁ : subgroup G) (H₂ : subgroup G) : subgroup G :=
```

Comment:
Should we turn this into an instance of `has_bracket`? See `src/algebra/lie/basic.lean`. If we do that, then we should factor the definition of `has_bracket` (+ notation) out of that file. Because this PR shouldn't depend on `lie.basic`.

Reply:
I was wondering this too, just now. We might have to be a little careful with instance priorities since for example two Lie ideals are subgroups of the Lie algebra but I think it would work, and would be worth having.

Reply:
I'm not sure if this is worth considering here but in the land Lie theory I set things up slightly more generally. I think think the equivalent here would be:
```
def really_general_commutator {G G' : Type*} [group G] [group G'] [mul_action G G']
  (H : subgroup G) (H' : subgroup G') : subgroup G' :=
subgroup.closure {x | ∃ (p ∈ H) (q ∈ H'), (p • q) * (p⁻¹ • q⁻¹) = x}
```

Reply:
Actually the right definition would probably be:
```lean
def really_general_commutator' {G G' : Type*} [monoid G] [group G'] [mul_action G G']
  (H : submonoid G) (H' : subgroup G') : subgroup G' :=
subgroup.closure {x | ∃ (p ∈ H) (q ∈ H'), (p • q) * q⁻¹ = x}
```
and then we'd recapture `general_commutator` using the self-action `p • q = p * q * p⁻¹`.

Reply:
@jcommelin I'm a little confused how `has_bracket` would be used here. In other cases it looks like `has_bracket` is for binary operations on elements. Or are you suggesting we should define the commutator of two elements of a group and then use that to define the commutator of two subgroups?

Reply:
@pglutz I think @jcommelin is thinking of the definition [here](https://github.com/leanprover-community/mathlib/blob/e35a703f907e325031dfcc78fcfe7a1b0456badf/src/algebra/lie/basic.lean#L930)

Reply:
@ocfnash I see, so `has_bracket` is used both for the operation on elements and the operation on ideals/etc. What is the value of `has_bracket`? Just that you get access to the notation? Or is there something else?

Reply:
No the value is just in getting the notation, which seems nice.

Reply:
Okay, that seems reasonable to me. Should that happen in this PR or in a future PR?

Reply:
And where should `has_bracket` go?

Reply:
I guess we can put the definition + notation of `has_bracket` in something like `data/bracket.lean`.

For me it's fine if you do everything in this PR.

Reply:
I've now added the file `data/bracket.lean` and switched to using that notation in `solvable.lean`. It seems to work fine though I had to add an extra coercion in a couple of places. Also, I haven't yet tried to modify the files in `algebra/lie`. Should I try to do that in this PR as well?

Reply:
nevermind about the last question. See below.